File: cpp/test/TestByteDecoder.cpp

#include "ParserSession.h"
#include "ByteDecoder.h"
#include "CodePoint.h"

#include "gtest/gtest.h"

class ParserSession;

using ParserSessionPtr = ParserSession *;


class ByteDecoderTest : public ::testing::Test {
protected:
    
    static ParserSessionPtr session;
    
    static void SetUpTestSuite() {
        
        session = new ParserSession();
    }
    
    static void TearDownTestSuite() {
        
        delete session;
    }
    
    void SetUp() override {
        
    }
    
    void TearDown() override {
        
        session->deinit();
    }
};

ParserSessionPtr ByteDecoderTest::session;


TEST_F(ByteDecoderTest, Basic1) {
    
    auto strIn = std::string("1+2");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('2'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

//
// UTF-8 bytes for \[Alpha] are {206, 177}
//
TEST_F(ByteDecoderTest, Basic2) {
    
    const unsigned char arr[] = {'1', '+', 206, 177};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 4, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(0x03b1));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    //
    // Issue: Non-ASCII character: ``"α" (\[Alpha])``
    //
    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
TEST_F(ByteDecoderTest, Basic3) {
    
    const unsigned char arr[] = {'1', '+', 0xE2, 0x9A, 0xA1};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 5, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(0x26A1));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    //
    // Issue: Non-ASCII character: ``"⚡" (\:26a1)``
    //
    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(ByteDecoderTest, Invalid1) {
    
    const unsigned char arr[] = {'1', '+', 0xf8};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 3, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_1_BYTE_UTF8_SEQUENCE));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

//
// UTF-8 bytes for \[Alpha] are {206, 177}
//
// So test with only first byte
//
TEST_F(ByteDecoderTest, Invalid2) {
    
    const unsigned char arr[] = {'1', '+', 206};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 3, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_1_BYTE_UTF8_SEQUENCE));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
TEST_F(ByteDecoderTest, Invalid3) {
    
    const unsigned char arr[] = {'1', '+', 0xE2};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 3, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xE2 byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_1_BYTE_UTF8_SEQUENCE));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
TEST_F(ByteDecoderTest, Invalid4) {
    
    const unsigned char arr[] = {'1', '+', 0xE2, 0x9A};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 4, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);

    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xE2 byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_2_BYTE_UTF8_SEQUENCE));

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

//
// Surrogates
//
TEST_F(ByteDecoderTest, Surrogate1) {
    
    const unsigned char arr[] = {'1', '+', 0xed, 0xa0, 0x80};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 5, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    EXPECT_EQ(session->buffer, session->start + 0);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    EXPECT_EQ(session->buffer, session->start + 1);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xED byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_3_BYTE_UTF8_SEQUENCE));

    EXPECT_EQ(session->buffer, session->start + 2);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xA0 byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->buffer, session->start + 5);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    //
    // Issue: Invalid UTF-8 sequence: Stray surrogate
    //
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

//
// Surrogates
//
TEST_F(ByteDecoderTest, Surrogate2) {
    
    const unsigned char arr[] = {'1', '+', 0xed, 0xb0, 0x80};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 5, nullptr, opts);

    auto c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('1'));

    EXPECT_EQ(session->buffer, session->start + 0);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    EXPECT_EQ(c, SourceCharacter('+'));

    EXPECT_EQ(session->buffer, session->start + 1);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xED byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_UNSAFE_3_BYTE_UTF8_SEQUENCE));

    EXPECT_EQ(session->buffer, session->start + 2);

    ByteDecoder_nextSourceCharacter(session, TOPLEVEL);
    
    c = ByteDecoder_currentSourceCharacter(session, TOPLEVEL);

    // from 0xB0 byte
    EXPECT_EQ(c, SourceCharacter(CODEPOINT_ENDOFFILE));

    EXPECT_EQ(session->buffer, session->start + 5);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    //
    // Issue: Invalid UTF-8 sequence: Stray surrogate
    //
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

File: crates/wolfram-parser/src/tests/test_byte_decoder.rs
use pretty_assertions::assert_eq;

use crate::{
    read::code_point::CodePoint, read::Reader, source::TOPLEVEL, ParseOptions,
};

#[test]
fn ByteDecoderTest_Empty() {
    let strIn = "";

    let mut reader = Reader::new(strIn.as_bytes(), &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 0);
}


#[test]
fn ByteDecoderTest_Basic1() {
    let strIn = "1+2";

    let mut reader = Reader::new(strIn.as_bytes(), &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '2');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 0);
}

//
// UTF-8 bytes for \[Alpha] are {206, 177}
//
#[test]
fn ByteDecoderTest_Basic2() {
    let arr = &[b'1', b'+', 206, 177];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c: CodePoint = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c.as_i32(), 0x03b1);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    //
    // Issue: Non-ASCII character: ``"α" (\[Alpha])``
    //
    assert_eq!(reader.non_fatal_issues.len(), 1);
    assert_eq!(reader.fatal_issues.len(), 0);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
#[test]
fn ByteDecoderTest_Basic3() {
    let arr = &[b'1', b'+', 0xE2, 0x9A, 0xA1];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c.as_i32(), 0x26A1);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    //
    // Issue: Non-ASCII character: ``"⚡" (\:26a1)``
    //
    assert_eq!(reader.non_fatal_issues.len(), 1);
    assert_eq!(reader.fatal_issues.len(), 0);
}

#[test]
fn ByteDecoderTest_Invalid1() {
    let arr = &[b'1', b'+', 0xf8];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::Unsafe1ByteUtf8Sequence);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 1);
}

//
// UTF-8 bytes for \[Alpha] are {206, 177}
//
// So test with only first byte
//
#[test]
fn ByteDecoderTest_Invalid2() {
    let arr = &[b'1', b'+', 206];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::Unsafe1ByteUtf8Sequence);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 1);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
#[test]
fn ByteDecoderTest_Invalid3() {
    let arr = &[b'1', b'+', 0xE2];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xE2 byte
    assert_eq!(c, CodePoint::Unsafe1ByteUtf8Sequence);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 1);
}

//
// UTF-8 bytes for HIGH VOLTAGE SIGN (U+26A1) are { 0xE2, 0x9A, 0xA1 }
//
#[test]
fn ByteDecoderTest_Invalid4() {
    let arr = &[b'1', b'+', 0xE2, 0x9A];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xE2 byte
    assert_eq!(c, CodePoint::Unsafe2ByteUtf8Sequence);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    assert_eq!(reader.fatal_issues.len(), 1);
}

//
// Surrogates
//
#[test]
fn ByteDecoderTest_Surrogate1() {
    let arr = &[b'1', b'+', 0xed, 0xa0, 0x80];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    assert_eq!(reader.buffer(), &arr[0..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    assert_eq!(reader.buffer(), &arr[1..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xED byte
    assert_eq!(c, CodePoint::Unsafe3ByteUtf8Sequence);

    assert_eq!(reader.buffer(), &arr[2..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xA0 byte
    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.buffer(), &arr[5..]);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    //
    // Issue: Invalid UTF-8 sequence: Stray surrogate
    //
    assert_eq!(reader.fatal_issues.len(), 1);
}

//
// Surrogates
//
#[test]
fn ByteDecoderTest_Surrogate2() {
    let arr = &[b'1', b'+', 0xed, 0xb0, 0x80];

    let mut reader = Reader::new(arr, &ParseOptions::default());

    let mut c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '1');

    assert_eq!(reader.buffer(), &arr[0..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    assert_eq!(c, '+');

    assert_eq!(reader.buffer(), &arr[1..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xED byte
    assert_eq!(c, CodePoint::Unsafe3ByteUtf8Sequence);

    assert_eq!(reader.buffer(), &arr[2..]);

    reader.next_source_char(TOPLEVEL);

    c = reader.peek_source_char(TOPLEVEL);

    // from 0xB0 byte
    assert_eq!(c, CodePoint::EndOfFile);

    assert_eq!(reader.buffer(), &arr[5..]);

    assert_eq!(reader.non_fatal_issues.len(), 0);
    //
    // Issue: Invalid UTF-8 sequence: Stray surrogate
    //
    assert_eq!(reader.fatal_issues.len(), 1);
}
